src/models/state/wallet/wallet_state.rs

```diff
@@ -3734,4 +3930,383 @@ mod tests {
+            let wallet_secret = WalletEntropy::new_pseudorandom(rng.random());
+            let data_dir = unit_test_data_directory(network).unwrap();
+            let wallet_state = WalletState::new_from_wallet_entropy(
+                &data_dir,
+                wallet_secret,
+                &cli_args::Args::default(),
+            )
+            .await;
+
+            // generate iterators for future keys
+            let generation_counter = wallet_state.wallet_db.get_generation_key_counter().await;
+            let symmetric_counter = wallet_state.wallet_db.get_symmetric_key_counter().await;
+
+            // don't just generate the iterators; run through them also
+            let num_future_keys = 100;
+            let future_generation_keys = wallet_state
+                .get_future_generation_spending_keys(num_future_keys)
+                .await
+                .collect_vec();
+            let future_symmetric_keys = wallet_state
+                .get_future_symmetric_keys(num_future_keys)
+                .await
+                .collect_vec();
+
+            // verify that the counters haven't changed
+            assert_eq!(
+                generation_counter,
+                wallet_state.wallet_db.get_generation_key_counter().await
+            );
+            assert_eq!(
+                symmetric_counter,
+                wallet_state.wallet_db.get_symmetric_key_counter().await
+            );
+
+            // make sure passing over the iterators is not being optimized away
+            black_box(future_generation_keys);
+            black_box(future_symmetric_keys);
+        }
+
+        /// Test that the method
+        /// [`WalletState::scan_for_utxos_announced_to_future_keys`] behaves as
+        /// exepected.
+        ///
+        /// Specifically:
+        ///  - Generate a random transaction kernel.
+        ///  - Generate future generation and symmetric keys, with random
+        ///    offsets.
+        ///  - Generate UTXOs with all necessary supplementary information, one
+        ///    for each key.
+        ///  - Generate public announcements for all UTXOs and put them into the
+        ///    transaction kernel.
+        ///  - For a subset of UTXOs, put the corresponding addition record into
+        ///    the transaction kernel.
+        ///  - Scan!
+        ///  - Verify that the UTXOs coming back from scanning matches with the
+        ///    master list of UTXOs filtered for:
+        ///     a) the addition record was selected for placement into the
+        ///        kernel;
+        ///     b) the relative index is smaller than num_future_keys.
+        ///
+        #[traced_test]
+        #[tokio::test]
+        async fn scan_for_utxos_announced_to_future_keys_behaves() {
+            let network = Network::Main;
+            let mut rng = StdRng::from_rng(&mut rng());
+            let wallet_secret = WalletEntropy::new_pseudorandom(rng.random());
```

Comment:
Are there advantages to generating a new rng this way? It should be possible to use `random()` or `rng().random_range()` directly, I think.

```suggestion
            let wallet_secret = WalletEntropy::new_pseudorandom(random());
```

Reply:
Yes, there are! If the test fails, it pays to have determinism to aid in debugging ;-)

Reply:
Then I'd suggest to debug-print the seed: in case the test is flaky, it's better to have a known bad case than it is to go hunting for one.

```suggestion
            let seed = random();
            dbg!(seed);
            let mut rng = StdRng::from_seed(seed);
            let wallet_secret = WalletEntropy::new_pseudorandom(rng.random());
```